File: DataSources/KiwoomApiSession.vb
Option Strict On
Option Explicit On
Option Infer Off

Imports System.Globalization
Imports System.Net
Imports System.Net.Http
Imports System.Text
Imports System.Text.Json
Imports System.Threading
Imports ChartKit.Core

Namespace DataSources
    Public Interface IKiwoomClock
        ReadOnly Property UtcNow As DateTimeOffset
        ReadOnly Property TickCount64 As Long
        Sub Sleep(milliseconds As Integer)
    End Interface

    Public NotInheritable Class SystemKiwoomClock
        Implements IKiwoomClock

        Public ReadOnly Property UtcNow As DateTimeOffset Implements IKiwoomClock.UtcNow
            Get
                Return DateTimeOffset.UtcNow
            End Get
        End Property

        Public ReadOnly Property TickCount64 As Long Implements IKiwoomClock.TickCount64
            Get
                Return Environment.TickCount64
            End Get
        End Property

        Public Sub Sleep(milliseconds As Integer) Implements IKiwoomClock.Sleep
            If milliseconds > 0 Then Thread.Sleep(milliseconds)
        End Sub
    End Class

    Public NotInheritable Class KiwoomApiSessionOptions
        Public Property RestHost As String = ""
        Public Property AppKey As String = ""
        Public Property SecretKey As String = ""
        Public Property IsMock As Boolean
        Public Property MinRequestIntervalMs As Integer
        Public Property MaxRateLimitRetries As Integer = 4
        Public Property TokenRefreshSkew As TimeSpan = TimeSpan.FromMinutes(5)
        Public Property TokenFallbackLifetime As TimeSpan = TimeSpan.FromHours(23)

        Public Sub Validate()
            If String.IsNullOrWhiteSpace(RestHost) Then
                Throw New ArgumentException("Kiwoom REST host가 비어 있습니다.", NameOf(RestHost))
            End If
            If String.IsNullOrWhiteSpace(AppKey) OrElse
               String.IsNullOrWhiteSpace(SecretKey) Then
                Throw New InvalidOperationException(
                    "키움 API 키 없음. .env 에 KIWOOM_APP_KEY / KIWOOM_SECRET_KEY (또는 REAL/MOCK) 설정 필요.")
            End If
            If MinRequestIntervalMs < 0 Then
                Throw New ArgumentOutOfRangeException(NameOf(MinRequestIntervalMs))
            End If
            If MaxRateLimitRetries < 0 Then
                Throw New ArgumentOutOfRangeException(NameOf(MaxRateLimitRetries))
            End If
            If TokenRefreshSkew < TimeSpan.Zero Then
                Throw New ArgumentOutOfRangeException(NameOf(TokenRefreshSkew))
            End If
            If TokenFallbackLifetime <= TimeSpan.Zero Then
                Throw New ArgumentOutOfRangeException(NameOf(TokenFallbackLifetime))
            End If
        End Sub
    End Class

    Public NotInheritable Class KiwoomApiSession
        Implements IDisposable

        Private ReadOnly _options As KiwoomApiSessionOptions
        Private ReadOnly _http As HttpClient
        Private ReadOnly _clock As IKiwoomClock
        Private ReadOnly _tokenSync As New Object()
        Private ReadOnly _rateSync As New Object()

        Private _token As String = ""
        Private _tokenExpiresUtc As DateTimeOffset = DateTimeOffset.MinValue
        Private _lastCallTicks As Long
        Private _blockedUntilTicks As Long
        Private _disposed As Boolean

        Public Sub New(options As KiwoomApiSessionOptions,
                       Optional handler As HttpMessageHandler = Nothing,
                       Optional clock As IKiwoomClock = Nothing)
            If options Is Nothing Then Throw New ArgumentNullException(NameOf(options))
            options.Validate()

            _options = options
            _clock = If(clock, New SystemKiwoomClock())
            _http = If(handler Is Nothing,
                       New HttpClient(),
                       New HttpClient(handler, disposeHandler:=True))
            _lastCallTicks = -CLng(_options.MinRequestIntervalMs)
        End Sub

        Public Function GetAccessToken() As String
            ThrowIfDisposed()

            SyncLock _tokenSync
                If IsTokenUsable() Then Return _token

                Dim issuedAt As DateTimeOffset = _clock.UtcNow
                Dim url As String = _options.RestHost.TrimEnd("/"c) & "/oauth2/token"
                Dim body As String = JsonSerializer.Serialize(
                    New Dictionary(Of String, String) From {
                        {"grant_type", "client_credentials"},
                        {"appkey", _options.AppKey},
                        {"secretkey", _options.SecretKey}})

                Using request As New HttpRequestMessage(HttpMethod.Post, url)
                    request.Content = New StringContent(body, Encoding.UTF8, "application/json")
                    Using response As HttpResponseMessage =
                        _http.SendAsync(request).GetAwaiter().GetResult()
                        Dim json As String =
                            response.Content.ReadAsStringAsync().GetAwaiter().GetResult()
                        If Not response.IsSuccessStatusCode Then
                            Throw New HttpRequestException(
                                $"Kiwoom token request failed: HTTP {CInt(response.StatusCode)} {response.ReasonPhrase}; body={json}",
                                Nothing,
                                response.StatusCode)
                        End If

                        Using document As JsonDocument = JsonDocument.Parse(json)
                            Dim root As JsonElement = document.RootElement
                            _token = ReadToken(root)
                            _tokenExpiresUtc = ResolveTokenExpiry(root, issuedAt)
                        End Using
                    End Using
                End Using

                If String.IsNullOrWhiteSpace(_token) Then
                    _tokenExpiresUtc = DateTimeOffset.MinValue
                    Throw New InvalidOperationException("토큰 발급 실패")
                End If

                Return _token
            End SyncLock
        End Function

        Public Sub InvalidateToken(Optional expectedToken As String = Nothing)
            SyncLock _tokenSync
                If Not String.IsNullOrEmpty(expectedToken) AndAlso
                   Not String.Equals(_token, expectedToken, StringComparison.Ordinal) Then
                    Return
                End If
                _token = ""
                _tokenExpiresUtc = DateTimeOffset.MinValue
            End SyncLock
        End Sub

        Public Function PostJson(path As String,
                                 apiId As String,
                                 body As String,
                                 contYn As String,
                                 nextKey As String,
                                 ByRef outCont As String,
                                 ByRef outNext As String) As JsonDocument
            ThrowIfDisposed()

            Dim url As String = _options.RestHost.TrimEnd("/"c) & path
            Dim rateRetryCount As Integer = 0
            Dim authenticationRetried As Boolean = False

            Do
                Dim accessToken As String = GetAccessToken()
                WaitForRequestTurn()

                Using request As New HttpRequestMessage(HttpMethod.Post, url)
                    request.Content = New StringContent(body, Encoding.UTF8, "application/json")
                    request.Headers.TryAddWithoutValidation("authorization", "Bearer " & accessToken)
                    request.Headers.TryAddWithoutValidation("api-id", apiId)
                    request.Headers.TryAddWithoutValidation("cont-yn", If(contYn, "N"))
                    request.Headers.TryAddWithoutValidation("next-key", If(nextKey, ""))

                    Using response As HttpResponseMessage =
                        _http.SendAsync(request).GetAwaiter().GetResult()
                        Dim json As String =
                            response.Content.ReadAsStringAsync().GetAwaiter().GetResult()

                        If (response.StatusCode = HttpStatusCode.Unauthorized OrElse
                            response.StatusCode = HttpStatusCode.Forbidden) AndAlso
                           Not authenticationRetried Then
                            authenticationRetried = True
                            InvalidateToken(accessToken)
                            Continue Do
                        End If

                        If response.StatusCode = HttpStatusCode.TooManyRequests AndAlso
                           rateRetryCount < _options.MaxRateLimitRetries Then
                            Dim retryMs As Integer = GetRetryDelayMs(response, rateRetryCount)
                            rateRetryCount += 1
                            RegisterGlobalBackoff(retryMs)
                            ChartLog.Info(
                                $"[KIWOOM] 429 api-id={apiId}, retry={rateRetryCount}/{_options.MaxRateLimitRetries}, wait={retryMs}ms")
                            Continue Do
                        End If

                        If Not response.IsSuccessStatusCode Then
                            Throw New HttpRequestException(
                                $"Kiwoom {apiId} failed: HTTP {CInt(response.StatusCode)} {response.ReasonPhrase}; body={json}",
                                Nothing,
                                response.StatusCode)
                        End If

                        outCont = HeaderOrDefault(response, "cont-yn", "N")
                        outNext = HeaderOrDefault(response, "next-key", "")
                        Return JsonDocument.Parse(json)
                    End Using
                End Using
            Loop
        End Function

        Private Function IsTokenUsable() As Boolean
            If String.IsNullOrWhiteSpace(_token) Then Return False
            Return _clock.UtcNow < (_tokenExpiresUtc - _options.TokenRefreshSkew)
        End Function

        Private Sub WaitForRequestTurn()
            SyncLock _rateSync
                Do
                    Dim nowTicks As Long = _clock.TickCount64
                    Dim intervalReady As Long =
                        _lastCallTicks + CLng(_options.MinRequestIntervalMs)
                    Dim readyAt As Long = Math.Max(intervalReady, _blockedUntilTicks)
                    Dim waitTicks As Long = readyAt - nowTicks
                    If waitTicks <= 0L Then Exit Do

                    Dim waitMs As Integer = CInt(Math.Min(CLng(Integer.MaxValue), waitTicks))
                    _clock.Sleep(waitMs)
                Loop

                _lastCallTicks = _clock.TickCount64
            End SyncLock
        End Sub

        Private Sub RegisterGlobalBackoff(milliseconds As Integer)
            If milliseconds <= 0 Then Return
            SyncLock _rateSync
                Dim blockedUntil As Long = _clock.TickCount64 + CLng(milliseconds)
                If blockedUntil > _blockedUntilTicks Then
                    _blockedUntilTicks = blockedUntil
                End If
            End SyncLock
        End Sub

        Private Function GetRetryDelayMs(response As HttpResponseMessage,
                                         attempt As Integer) As Integer
            If response.Headers.RetryAfter IsNot Nothing Then
                If response.Headers.RetryAfter.Delta.HasValue Then
                    Return Math.Max(
                        _options.MinRequestIntervalMs,
                        CInt(Math.Ceiling(
                            response.Headers.RetryAfter.Delta.Value.TotalMilliseconds)))
                End If
                If response.Headers.RetryAfter.Date.HasValue Then
                    Dim milliseconds As Double =
                        (response.Headers.RetryAfter.Date.Value - _clock.UtcNow).TotalMilliseconds
                    If milliseconds > 0.0R Then
                        Return Math.Max(
                            _options.MinRequestIntervalMs,
                            CInt(Math.Ceiling(milliseconds)))
                    End If
                End If
            End If

            Dim baseDelay As Integer = Math.Max(1100, _options.MinRequestIntervalMs)
            Dim delayValue As Long = CLng(baseDelay) * CLng(attempt + 1)
            Return CInt(Math.Min(CLng(Integer.MaxValue), delayValue))
        End Function

        Private Function ResolveTokenExpiry(root As JsonElement,
                                            issuedAt As DateTimeOffset) As DateTimeOffset
            Dim expiresIn As JsonElement
            If root.TryGetProperty("expires_in", expiresIn) Then
                Dim seconds As Double
                If expiresIn.ValueKind = JsonValueKind.Number AndAlso
                   expiresIn.TryGetDouble(seconds) AndAlso seconds > 0.0R Then
                    Return issuedAt.AddSeconds(seconds)
                End If
                If expiresIn.ValueKind = JsonValueKind.String AndAlso
                   Double.TryParse(
                       expiresIn.GetString(),
                       NumberStyles.Float,
                       CultureInfo.InvariantCulture,
                       seconds) AndAlso seconds > 0.0R Then
                    Return issuedAt.AddSeconds(seconds)
                End If
            End If

            Dim expiryKeys As String() = {"expires_dt", "expires_at", "expiration"}
            For Each key As String In expiryKeys
                Dim expiryElement As JsonElement
                If Not root.TryGetProperty(key, expiryElement) OrElse
                   expiryElement.ValueKind <> JsonValueKind.String Then Continue For

                Dim parsed As DateTimeOffset
                If DateTimeOffset.TryParse(
                    expiryElement.GetString(),
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.AssumeLocal,
                    parsed) Then
                    Return parsed.ToUniversalTime()
                End If

                Dim localDate As DateTime
                If DateTime.TryParseExact(
                    expiryElement.GetString(),
                    New String() {"yyyyMMddHHmmss", "yyyy-MM-dd HH:mm:ss"},
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.None,
                    localDate) Then
                    Return New DateTimeOffset(
                        localDate,
                        TimeZoneInfo.Local.GetUtcOffset(localDate)).ToUniversalTime()
                End If
            Next

            Return issuedAt.Add(_options.TokenFallbackLifetime)
        End Function

        Private Shared Function ReadToken(root As JsonElement) As String
            Dim element As JsonElement
            If root.TryGetProperty("token", element) Then
                Return If(element.GetString(), "")
            End If
            If root.TryGetProperty("access_token", element) Then
                Return If(element.GetString(), "")
            End If
            Return ""
        End Function

        Private Shared Function HeaderOrDefault(response As HttpResponseMessage,
                                                key As String,
                                                defaultValue As String) As String
            Dim values As IEnumerable(Of String) = Nothing
            If response.Headers.TryGetValues(key, values) Then
                For Each value As String In values
                    Return value
                Next
            End If
            Return defaultValue
        End Function

        Private Sub ThrowIfDisposed()
            If _disposed Then Throw New ObjectDisposedException(NameOf(KiwoomApiSession))
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            If _disposed Then Return
            _disposed = True
            _http.Dispose()
        End Sub
    End Class

    Public NotInheritable Class KiwoomApiSessionProvider
        Private Shared ReadOnly DefaultSession As New Lazy(Of KiwoomApiSession)(
            AddressOf CreateDefaultSession,
            LazyThreadSafetyMode.ExecutionAndPublication)

        Private Sub New()
        End Sub

        Public Shared Function GetDefault() As KiwoomApiSession
            Return DefaultSession.Value
        End Function

        Private Shared Function CreateDefaultSession() As KiwoomApiSession
            Dim options As New KiwoomApiSessionOptions With {
                .RestHost = EnvConfig.RestHost,
                .AppKey = EnvConfig.AppKey,
                .SecretKey = EnvConfig.SecretKey,
                .IsMock = EnvConfig.IsMock,
                .MinRequestIntervalMs = If(EnvConfig.IsMock, 1100, 220),
                .MaxRateLimitRetries = 4
            }
            Return New KiwoomApiSession(options)
        End Function
    End Class
End Namespace

File: csharp_chartkit/src/ChartKit.DataSources/KiwoomApiSession.cs
using System.Globalization;
using System.Net;
using System.Text;
using System.Text.Json;

namespace ChartKit.CSharp.DataSources;

public sealed class KiwoomApiSession : IAsyncDisposable
{
    private static readonly TimeSpan TokenRefreshSkew = TimeSpan.FromMinutes(1);
    private readonly KiwoomOptions _options;
    private readonly HttpClient _http;
    private readonly bool _ownsHttpClient;
    private readonly IKiwoomClock _clock;
    private readonly SemaphoreSlim _tokenGate = new(1, 1);
    private readonly SemaphoreSlim _rateGate = new(1, 1);
    private readonly object _tokenSync = new();
    private string _accessToken = "";
    private DateTimeOffset _tokenExpiresAtUtc = DateTimeOffset.MinValue;
    private long _nextRequestTimestamp;
    private long _blockedUntilTimestamp;
    private int _disposed;

    public KiwoomApiSession(
        KiwoomOptions options,
        HttpMessageHandler? handler = null,
        IKiwoomClock? clock = null)
    {
        _options = options ?? throw new ArgumentNullException(nameof(options));
        _clock = clock ?? new SystemKiwoomClock();
        _ownsHttpClient = true;
        _http = handler is null
            ? new HttpClient()
            : new HttpClient(handler, disposeHandler: true);
        _http.Timeout = _options.RequestTimeout;
    }

    public KiwoomOptions Options => _options;

    public async Task<string> GetAccessTokenAsync(
        CancellationToken cancellationToken = default)
    {
        ThrowIfDisposed();
        if (TryGetUsableToken(out string cached)) return cached;

        await _tokenGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            if (TryGetUsableToken(out cached)) return cached;
            _options.ValidateCredentials();

            using var request = new HttpRequestMessage(
                HttpMethod.Post,
                new Uri(_options.RestBaseUri, "/oauth2/token"));
            string body = JsonSerializer.Serialize(new Dictionary<string, string>
            {
                ["grant_type"] = "client_credentials",
                ["appkey"] = _options.AppKey,
                ["secretkey"] = _options.SecretKey
            });
            request.Content = new StringContent(
                body,
                Encoding.UTF8,
                "application/json");

            using HttpResponseMessage response =
                await _http.SendAsync(request, cancellationToken).ConfigureAwait(false);
            byte[] payload =
                await response.Content.ReadAsByteArrayAsync(cancellationToken)
                    .ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
                throw CreateHttpException(response.StatusCode, payload, "token");

            using JsonDocument document = JsonDocument.Parse(payload);
            JsonElement root = document.RootElement;
            string token = ReadString(root, "token");
            if (token.Length == 0) token = ReadString(root, "access_token");
            if (token.Length == 0)
                throw new InvalidOperationException(
                    "Kiwoom token response did not contain a token.");

            lock (_tokenSync)
            {
                _accessToken = token;
                _tokenExpiresAtUtc = ResolveTokenExpiry(root, _clock.UtcNow);
            }
            return token;
        }
        finally
        {
            _tokenGate.Release();
        }
    }

    public void InvalidateToken(string exactToken)
    {
        if (string.IsNullOrEmpty(exactToken)) return;
        lock (_tokenSync)
        {
            if (!string.Equals(
                    _accessToken,
                    exactToken,
                    StringComparison.Ordinal))
                return;
            _accessToken = "";
            _tokenExpiresAtUtc = DateTimeOffset.MinValue;
        }
    }

    public async Task<KiwoomJsonResponse> PostJsonAsync(
        string path,
        string apiId,
        string jsonBody,
        string continuation = "N",
        string nextKey = "",
        CancellationToken cancellationToken = default)
    {
        ThrowIfDisposed();
        if (string.IsNullOrWhiteSpace(path))
            throw new ArgumentException("Path is required.", nameof(path));
        if (string.IsNullOrWhiteSpace(apiId))
            throw new ArgumentException("API id is required.", nameof(apiId));

        int authenticationRetries = 0;
        int throttlingRetries = 0;
        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();
            string token = await GetAccessTokenAsync(cancellationToken)
                .ConfigureAwait(false);
            using var request = new HttpRequestMessage(
                HttpMethod.Post,
                new Uri(_options.RestBaseUri, path));
            request.Headers.TryAddWithoutValidation(
                "authorization", "Bearer " + token);
            request.Headers.TryAddWithoutValidation("api-id", apiId);
            request.Headers.TryAddWithoutValidation("cont-yn", continuation);
            request.Headers.TryAddWithoutValidation("next-key", nextKey);
            request.Content = new StringContent(
                jsonBody,
                Encoding.UTF8,
                "application/json");

            using HttpResponseMessage response =
                await SendThrottledAsync(request, cancellationToken)
                    .ConfigureAwait(false);

            if (response.StatusCode is
                HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                InvalidateToken(token);
                if (authenticationRetries++ == 0) continue;
            }

            if ((int)response.StatusCode == 429 && throttlingRetries++ < 3)
            {
                TimeSpan delay = ResolveRetryAfter(response, _clock.UtcNow);
                SetGlobalBlock(delay);
                await _clock.DelayAsync(delay, cancellationToken)
                    .ConfigureAwait(false);
                continue;
            }

            byte[] payload =
                await response.Content.ReadAsByteArrayAsync(cancellationToken)
                    .ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
                throw CreateHttpException(response.StatusCode, payload, apiId);

            JsonDocument document = JsonDocument.Parse(payload);
            return new KiwoomJsonResponse(
                document,
                response.StatusCode,
                ReadHeader(response, "cont-yn", "N"),
                ReadHeader(response, "next-key", ""));
        }
    }

    private async Task<HttpResponseMessage> SendThrottledAsync(
        HttpRequestMessage request,
        CancellationToken cancellationToken)
    {
        Task<HttpResponseMessage> pending;
        await _rateGate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            long now = _clock.Timestamp;
            long target = Math.Max(
                Volatile.Read(ref _nextRequestTimestamp),
                Volatile.Read(ref _blockedUntilTimestamp));
            if (target > now)
                await _clock.DelayAsync(
                        ToTimeSpan(target - now),
                        cancellationToken)
                    .ConfigureAwait(false);

            long intervalTicks = ToTimestampTicks(_options.RequestInterval);
            Volatile.Write(
                ref _nextRequestTimestamp,
                Math.Max(_clock.Timestamp, target) + intervalTicks);

            pending = _http.SendAsync(
                request,
                HttpCompletionOption.ResponseHeadersRead,
                cancellationToken);
        }
        finally
        {
            _rateGate.Release();
        }

        return await pending.ConfigureAwait(false);
    }

    private bool TryGetUsableToken(out string token)
    {
        lock (_tokenSync)
        {
            token = _accessToken;
            return token.Length > 0 &&
                   _clock.UtcNow + TokenRefreshSkew < _tokenExpiresAtUtc;
        }
    }

    private void SetGlobalBlock(TimeSpan delay)
    {
        long candidate = _clock.Timestamp + ToTimestampTicks(delay);
        long current;
        do
        {
            current = Volatile.Read(ref _blockedUntilTimestamp);
            if (candidate <= current) return;
        } while (Interlocked.CompareExchange(
                     ref _blockedUntilTimestamp,
                     candidate,
                     current) != current);
    }

    private long ToTimestampTicks(TimeSpan value) =>
        Math.Max(0L,
            (long)Math.Ceiling(value.TotalSeconds * _clock.Frequency));

    private TimeSpan ToTimeSpan(long timestampTicks) =>
        TimeSpan.FromSeconds((double)timestampTicks / _clock.Frequency);

    private static TimeSpan ResolveRetryAfter(
        HttpResponseMessage response,
        DateTimeOffset now)
    {
        if (response.Headers.RetryAfter?.Delta is { } delta &&
            delta > TimeSpan.Zero)
            return delta;
        if (response.Headers.RetryAfter?.Date is { } date)
        {
            TimeSpan calculated = date - now;
            if (calculated > TimeSpan.Zero) return calculated;
        }
        return TimeSpan.FromSeconds(1);
    }

    private static DateTimeOffset ResolveTokenExpiry(
        JsonElement root,
        DateTimeOffset now)
    {
        if (TryReadDouble(root, "expires_in", out double seconds) &&
            seconds > 0)
            return now.AddSeconds(seconds);

        foreach (string key in new[]
                 {
                     "expires_dt", "expires_at", "expiration"
                 })
        {
            string value = ReadString(root, key);
            if (value.Length == 0) continue;
            if (DateTimeOffset.TryParse(
                    value,
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.AssumeUniversal,
                    out DateTimeOffset offset))
                return offset.ToUniversalTime();
            if (DateTime.TryParseExact(
                    value,
                    new[]
                    {
                        "yyyyMMddHHmmss", "yyyy-MM-dd HH:mm:ss"
                    },
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.None,
                    out DateTime local))
                return new DateTimeOffset(local).ToUniversalTime();
        }

        return now.AddMinutes(30);
    }

    private static bool TryReadDouble(
        JsonElement root,
        string key,
        out double value)
    {
        value = 0;
        if (!root.TryGetProperty(key, out JsonElement element)) return false;
        if (element.ValueKind == JsonValueKind.Number)
            return element.TryGetDouble(out value);
        return element.ValueKind == JsonValueKind.String &&
               double.TryParse(
                   element.GetString(),
                   NumberStyles.Any,
                   CultureInfo.InvariantCulture,
                   out value);
    }

    private static string ReadString(JsonElement root, string key)
    {
        if (!root.TryGetProperty(key, out JsonElement element)) return "";
        return element.ValueKind == JsonValueKind.String
            ? element.GetString() ?? ""
            : element.ToString();
    }

    private static string ReadHeader(
        HttpResponseMessage response,
        string name,
        string fallback)
    {
        if (response.Headers.TryGetValues(
                name,
                out IEnumerable<string>? values))
            return values.FirstOrDefault()?.Trim() ?? fallback;
        if (response.Content.Headers.TryGetValues(name, out values))
            return values.FirstOrDefault()?.Trim() ?? fallback;
        return fallback;
    }

    private static HttpRequestException CreateHttpException(
        HttpStatusCode statusCode,
        byte[] payload,
        string operation)
    {
        string text = Encoding.UTF8.GetString(payload);
        if (text.Length > 512) text = text[..512];
        return new HttpRequestException(
            $"Kiwoom {operation} failed with {(int)statusCode}: {text}",
            null,
            statusCode);
    }

    private void ThrowIfDisposed()
    {
        if (Volatile.Read(ref _disposed) != 0)
            throw new ObjectDisposedException(nameof(KiwoomApiSession));
    }

    public ValueTask DisposeAsync()
    {
        if (Interlocked.Exchange(ref _disposed, 1) != 0)
            return ValueTask.CompletedTask;
        _tokenGate.Dispose();
        _rateGate.Dispose();
        if (_ownsHttpClient) _http.Dispose();
        return ValueTask.CompletedTask;
    }
}
